network_test_harness.go

```diff
@@ -0,0 +1,302 @@
+package basicnet
+
+import (
+	"context"
+        "flag"
+	"fmt"
+	net "gx/ipfs/QmNa31VPzC561NWwRsJLE7nGYZYuuD2QfpK2b1q9BK54J1/go-libp2p-net"
+	peer "gx/ipfs/QmXYjuNuxVzXKJCfWasQk1RqkhVLDM9jtUKhqc2WPQmFSB/go-libp2p-peer"
+	crypto "gx/ipfs/QmaPbCnUMBohSGo3KnxEa2bHqyJVVeEEcwtqJAYxerieBo/go-libp2p-crypto"
+	"sync"
+	"testing"
+	"time"
+	"github.com/ericxtang/m3u8"
+	"github.com/golang/glog"
+)
+var nodeCount = flag.Int("nodeNum", 10, "number of nodes") 
+var networkTopology string
+var msgType string
+var rc = make(chan map[string]string)
+var finishResult  FinishStreamMsg
+
+
+func init() {
+     flag.StringVar(&networkTopology,"topo","star","network topology") 
+     flag.StringVar(&msgType, "msg","Sub","message type")
+     flag.Lookup("logtostderr").Value.Set("true")
+     var logLevel string
+     flag.StringVar(&logLevel, "logLevel", "3", "test")
+     flag.Lookup("v").Value.Set(logLevel)
+}
+
+func createNetwork(n int, networkTopo string) ([]*NetworkNode, []*BasicVideoNetwork) {
+    nodes := make([]*NetworkNode,n,n) 
+    vn := make([]*BasicVideoNetwork, n,n)
+    for  i:= 0; i < n ; i++ {
+	priv, pub, _ := crypto.GenerateKeyPair(crypto.RSA, 2048)
+	nodes[i], _ = NewNode(15000+i, priv, pub, &BasicNotifiee{})
+	vn[i],_ = NewBasicVideoNetwork(nodes[i], "")
+	if err := vn[i].SetupProtocol(); err != nil {
+		glog.Errorf("Error creating node: %v", err)
+	}
+    }
+    for  i:= 0; i < n ; i++ {
+        switch networkTopo {
+            case "ring":
+		      connectHosts(vn[i].NetworkNode.PeerHost, vn[(i+1)%n].NetworkNode.PeerHost)
+	    //default topology is star 
+            default:
+               if i > 0 {  
+		       connectHosts(vn[i].NetworkNode.PeerHost, vn[0].NetworkNode.PeerHost)
+	       }
+        }
+    }
+    return nodes, vn 
+
+}
+
+/*
+Create a network of N nodes(N>2) with 2 different topology options: star and ring, test sending and receiving different types of messages between random nodes and check each node's status after messages are sent
+To run the test:
+go test -run SendNetworkMsg -args  -topo ring  -msg TranscodeResponse  -nodeNum 5
+*/
+
```

Comment:
I think we'll want the default `go test` to skips the integration tests.  Can we:
1. Trigger the integration tests with `go test -integration`.
2. Enable this by adding a `fund init()` that checks if the `-integration` flag is passed in, if not, set a `skip` flag and do `t.Skip`.
3. Instead of params from the command line, make the params function parameters.  So we'll have something like `runNetworkIntegrationTest(param1, param2, ...)`, and be able to run multiple integration tests in the same run by calling it.

Reply:
OK good advice, For 3 "be able to run multiple integration tests in the same run by calling it" do you mean running the runNetworkIntegrationTest in a wrapper Test function?

Reply:
Yes.  For example, this way we can run tests for ring-shaped network and star-shaped network in the same run.